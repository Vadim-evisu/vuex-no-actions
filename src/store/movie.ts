import { Module } from 'vuex';

export interface IMovieStoreState {
  isLoadInProgress: boolean;
  data: any;
}

export const mutationNames = {
  loadStarted: 'loadStarted',
  loadCompleted: 'loadCompleted',
};

export const createInitialState = (): IMovieStoreState => ({
  isLoadInProgress: false,
  data: {},
});

function createModule(): Module<IMovieStoreState, any> {
  const module = {
    namespaced: true,
    state: () => createInitialState(),
    mutations: {
      [mutationNames.loadStarted](state: IMovieStoreState): void {
        state.isLoadInProgress = true;
      },
      [mutationNames.loadCompleted](
        state: IMovieStoreState,
        payload: any
      ): void {
        state.isLoadInProgress = false;
        state.data = payload.data;
      },
    },
  };
  return module;
}

export { createModule };
