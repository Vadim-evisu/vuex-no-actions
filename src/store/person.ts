import { Module } from 'vuex';

export interface IPersonStoreState {
  isLoadInProgress: boolean;
  data: any;
}

export const mutationNames = {
  loadStarted: 'loadStarted',
  loadCompleted: 'loadCompleted',
};

export const createInitialState = (): IPersonStoreState => ({
  isLoadInProgress: false,
  data: {},
});

function createModule(): Module<IPersonStoreState, any> {
  const module = {
    namespaced: true,
    state: () => createInitialState(),
    mutations: {
      [mutationNames.loadStarted](state: IPersonStoreState): void {
        state.isLoadInProgress = true;
      },
      [mutationNames.loadCompleted](
        state: IPersonStoreState,
        payload: any
      ): void {
        state.isLoadInProgress = false;
        state.data = payload.data;
      },
    },
  };
  return module;
}

export { createModule };
