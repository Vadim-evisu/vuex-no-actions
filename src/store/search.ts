import { Module } from 'vuex';
export interface ISearchStoreState {
  isSearchInProgress: boolean;
  searchResults: any;
}

export const getterNames = {
  resultsCount: 'resultsCount',
};

export const mutationNames = {
  searchStarted: 'searchStarted',
  searchCompleted: 'searchCompleted',
};

export const createInitialState = (): ISearchStoreState => ({
  isSearchInProgress: false,
  searchResults: {},
});

function createModule(): Module<ISearchStoreState, any> {
  const module = {
    namespaced: true,
    state: () => createInitialState(),
    getters: {
      [getterNames.resultsCount]: (state: ISearchStoreState) =>
        (state.searchResults.results || []).length,
    },
    mutations: {
      [mutationNames.searchStarted](state: ISearchStoreState): void {
        state.isSearchInProgress = true;
        state.searchResults = {};
      },
      [mutationNames.searchCompleted](
        state: ISearchStoreState,
        payload: { results: any }
      ): void {
        state.isSearchInProgress = false;
        state.searchResults = payload.results;
      },
    },
  };
  return module;
}

export { createModule };
