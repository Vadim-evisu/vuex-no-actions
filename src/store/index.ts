import Vuex, { Store } from 'vuex';
import * as searchModule from './search';
import * as movieModule from './movie';
import * as personModule from './person';

export interface IRootState {}

const moduleNames = {
  search: 'search',
  movie: 'movie',
  person: 'person',
};

function createStore(): Store<IRootState> {
  const store = new Vuex.Store({
    state: {},
    mutations: {},
    actions: {},
  });

  store.registerModule(moduleNames.search, searchModule.createModule());
  store.registerModule(moduleNames.movie, movieModule.createModule());
  store.registerModule(moduleNames.person, personModule.createModule());
  return store;
}

export { moduleNames, createStore };
