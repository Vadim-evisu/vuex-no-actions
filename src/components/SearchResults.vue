<template>
  <div>
    <progress max="100" v-show="isSearchInProgress"></progress>
    <div v-show="resultsCount === 0 && !isSearchInProgress">
      go ahead and make a search
    </div>
    <div data-test-id="searchResults">
      <div v-for="item in searchResults" v-bind:key="item.id">
        <ResultItemDisplayCard :item="item" />
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import Vue from 'vue';
import { mapGetters, mapState } from 'vuex';
import { ISearchStoreState, getterNames } from '@/store/search';
import { moduleNames } from '@/store';
import ResultItemDisplayCard from './ResultItemDisplayCard.vue';
import * as searchService from '@/services/searchService';

export default Vue.extend({
  async mounted() {
    await this.searchAsync();
  },
  components: {
    ResultItemDisplayCard,
  },
  computed: {
    ...mapState<ISearchStoreState>(moduleNames.search, {
      searchResults: (state: ISearchStoreState) => state.searchResults.results,
      isSearchInProgress: (state: ISearchStoreState) =>
        state.isSearchInProgress,
    }),
    ...mapGetters(moduleNames.search, [getterNames.resultsCount]),
  },
  props: {
    query: String,
  },
  watch: {
    query: 'searchAsync',
  },
  methods: {
    async searchAsync() {
      await searchService.searchAsync(this.$store.commit, this.query);
    },
  },
});
</script>
