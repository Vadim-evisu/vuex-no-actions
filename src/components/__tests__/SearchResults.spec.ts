import { createLocalVue, shallowMount, Wrapper } from '@vue/test-utils';
import Vuex from 'vuex';
import { createInitialState } from '@/store/search';

import SearchResults from '../SearchResults.vue';
import { moduleNames } from '@/store';
import * as sampleApiResponse from './seach-multi-sample-response.json';

describe('when mounted', () => {
  it('tests work', () => {
    expect(true).toBeTruthy();
  });

  const localVue = createLocalVue();
  localVue.use(Vuex);

  const searchModule = {
    namespaced: true,
    state: () => createInitialState(),
  };
  let wrapper: Wrapper<any>;

  const testQuery = `test query ${Date.now()}`;

  beforeEach(() => {
    const store = new Vuex.Store({});
    store.registerModule(moduleNames.search, searchModule);

    wrapper = shallowMount(SearchResults, {
      propsData: {
        query: testQuery,
      },
      localVue,
      store,
    });
  });
});

describe('when search is in progress', () => {
  const localVue = createLocalVue();
  localVue.use(Vuex);

  const searchModule = {
    namespaced: true,
    state: () => ({
      ...createInitialState(),
      isSearchInProgress: true,
    }),
  };
  let wrapper: Wrapper<any>;

  beforeEach(() => {
    const store = new Vuex.Store({});
    store.registerModule(moduleNames.search, searchModule);

    wrapper = shallowMount(SearchResults, {
      localVue,
      store,
    });
  });

  it(`displays the search in progress text`, () => {
    expect(wrapper).toMatchSnapshot();
  });
});

describe('when there are search results', () => {
  const localVue = createLocalVue();
  localVue.use(Vuex);

  const searchModule = {
    namespaced: true,
    state: () => ({
      ...createInitialState(),
      isSearchInProgress: false,
      searchResults: sampleApiResponse,
    }),
  };
  let wrapper: Wrapper<any>;

  beforeEach(() => {
    const store = new Vuex.Store({});
    store.registerModule(moduleNames.search, searchModule);

    wrapper = shallowMount(SearchResults, {
      localVue,
      store,
    });
  });

  it(`displays the search results`, () => {
    expect(wrapper).toMatchSnapshot();
  });
});
