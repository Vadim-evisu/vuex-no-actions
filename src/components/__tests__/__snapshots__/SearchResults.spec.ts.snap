// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when search is in progress displays the search in progress text 1`] = `
<div><progress max="100"></progress>
  <div style="display: none;">
    go ahead and make a search
  </div>
  <div data-test-id="searchResults"></div>
</div>
`;

exports[`when there are search results displays the search results 1`] = `
<div><progress max="100" style="display: none;"></progress>
  <div style="display: none;">
    go ahead and make a search
  </div>
  <div data-test-id="searchResults">
    <div>
      <resultitemdisplaycard-stub item="[object Object]"></resultitemdisplaycard-stub>
    </div>
    <div>
      <resultitemdisplaycard-stub item="[object Object]"></resultitemdisplaycard-stub>
    </div>
    <div>
      <resultitemdisplaycard-stub item="[object Object]"></resultitemdisplaycard-stub>
    </div>
    <div>
      <resultitemdisplaycard-stub item="[object Object]"></resultitemdisplaycard-stub>
    </div>
    <div>
      <resultitemdisplaycard-stub item="[object Object]"></resultitemdisplaycard-stub>
    </div>
    <div>
      <resultitemdisplaycard-stub item="[object Object]"></resultitemdisplaycard-stub>
    </div>
    <div>
      <resultitemdisplaycard-stub item="[object Object]"></resultitemdisplaycard-stub>
    </div>
    <div>
      <resultitemdisplaycard-stub item="[object Object]"></resultitemdisplaycard-stub>
    </div>
    <div>
      <resultitemdisplaycard-stub item="[object Object]"></resultitemdisplaycard-stub>
    </div>
    <div>
      <resultitemdisplaycard-stub item="[object Object]"></resultitemdisplaycard-stub>
    </div>
    <div>
      <resultitemdisplaycard-stub item="[object Object]"></resultitemdisplaycard-stub>
    </div>
    <div>
      <resultitemdisplaycard-stub item="[object Object]"></resultitemdisplaycard-stub>
    </div>
    <div>
      <resultitemdisplaycard-stub item="[object Object]"></resultitemdisplaycard-stub>
    </div>
    <div>
      <resultitemdisplaycard-stub item="[object Object]"></resultitemdisplaycard-stub>
    </div>
    <div>
      <resultitemdisplaycard-stub item="[object Object]"></resultitemdisplaycard-stub>
    </div>
    <div>
      <resultitemdisplaycard-stub item="[object Object]"></resultitemdisplaycard-stub>
    </div>
    <div>
      <resultitemdisplaycard-stub item="[object Object]"></resultitemdisplaycard-stub>
    </div>
    <div>
      <resultitemdisplaycard-stub item="[object Object]"></resultitemdisplaycard-stub>
    </div>
    <div>
      <resultitemdisplaycard-stub item="[object Object]"></resultitemdisplaycard-stub>
    </div>
    <div>
      <resultitemdisplaycard-stub item="[object Object]"></resultitemdisplaycard-stub>
    </div>
  </div>
</div>
`;
